### PR TITLE
HexEditor: Change cursor type from blinking caret to black box

### PR DESCRIPTION
**Pull request by tetektoza**
Discussion regarding this change was moved in this PR: https://github.com/SerenityOS/serenity/pull/22211, and also on #Userland channel, on serenity's discord.

These series of patches change cursor type from caret to black box for both Hex and Text modes, because right now the way how blinking caret looks like is more closer to "insert" mode in similar editors, whereas the real behavior of this cursor is more of a "replace" mode seen in similar editors like GHex.

There have also been fixed crashes when selecting last bytes with mouse and then writing something. Now the cursor is being moved to the last byte of selection, instead of moving the cursor to the byte one after last selected byte.

### CODE REVIEW

File: Userland/Applications/HexEditor/HexEditor.cpp [hunk 1]
```diff
@@ -421,6 +421,7 @@ void HexEditor::keydown_event(GUI::KeyEvent& event)
                 swap(m_selection_start, m_selection_end);
         } else
             m_selection_start = m_selection_end = m_position = new_position;
+
```
Comment (LucasChollet) [inline Userland/Applications/HexEditor/HexEditor.cpp:424]: Please avoid doing a formatting change if you're not modifying something close.
As this is improperly formatted regarding our rules, you can put that in its own commit if you want. (The rule says that if one branch require brackets, all branches should have some).

File: Userland/Applications/HexEditor/HexEditor.cpp [hunk 2]
```diff
@@ -716,16 +732,23 @@ void HexEditor::paint_event(GUI::PaintEvent& event)
...
                     };
-                    painter.fill_rect(cursor_position_rect, palette().text_cursor());
+                    painter.fill_rect(cursor_position_rect, palette().black());
                 }
             }
+
+            if (byte_position == m_position)
+                painter.draw_text(text_display_rect, character, Gfx::TextAlignment::TopLeft, palette().selection_text());
+            else
+                painter.draw_text(text_display_rect, character, Gfx::TextAlignment::TopLeft, text_color);
```
Comment (LucasChollet) [inline Userland/Applications/HexEditor/HexEditor.cpp]: This block looks really the one above, what are your thought about factorizing them?

**Review by LucasChollet — COMMENTED** [review]
I really like the direction this is taking.
There is one regression that I noticed though, sometimes one byte disappear when you're in the middle of the modification.
Screenshot from 2023-12-16 17-22-32
It does not happen every time, and I only achieved to trigger it on the two first bytes.

File: Userland/Applications/HexEditor/HexEditor.cpp (hunk 2 above)
Comment (tetektoza) [reply Userland/Applications/HexEditor/HexEditor.cpp]: I refactored double initialization of background/text colors, you can take a peek in the new commit, IMO looks a little bit clearer but I'm not sure what you think. This code that you've pointed out above has changed a bit since I've addressed your report about a regression, but I've tried to move this chunk in lines 736-746 to a separate lamba as we've discussed on discord, which would remove some repetition, but I don't know why (as also discussed on discord) it seems that when we do that, black box cursor gets drawn on top of the text after this modification. It's weird, since text is being drawn AFTER the call to lambda. I can debug a little bit further if you want, if you'd care to help I'd gladly take your assistance, but I couldn't wrap my head why it happens.

+ also added one more commit since I've found (and fixed) a bug where you could go outside document's boundaries with GoToOffsetDialog.

File: Userland/Applications/HexEditor/GoToOffsetDialog.cpp [hunk 3]
```diff
@@ -73,8 +73,8 @@ int GoToOffsetDialog::calculate_new_offset(int input_offset)
         VERIFY_NOT_REACHED();
     }
 
-    if (new_offset > m_buffer_size)
-        new_offset = m_buffer_size;
+    if (new_offset > m_buffer_size - 1)
```
Comment (LucasChollet) [inline Userland/Applications/HexEditor/GoToOffsetDialog.cpp:76]: Please avoid putting unrelated stuff in your PR, that's never going to help you get them merged earlier.

This problem is (unsurprisingly) quite recurrent, so let's try to factorize every modification made to the cursor into one function, so every single check (and bug :)) is located at the same place. But yeah, that looks like work for another PR.

**Review by LucasChollet — COMMENTED** [review]
Except the last commit, that looks mergeable to me.
I've tested it again and everything looks fine :+1:
